### ISSUE TITLE
Query: compilation error (bad cancellation token) for async query projecting collection navigation and include

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#12463]
query using complex navs model:

```cs
from l1 in ctx.LevelOnes.Include(l => l.OneToMany_Optional)
select new
{
    l1,
    l1.OneToMany_Optional
},
```

exception:
```
System.InvalidOperationException : variable 'ct' of type 'System.Threading.CancellationToken' referenced from scope '', but it is not defined
	at System.Linq.Expressions.Compiler.VariableBinder.Reference(ParameterExpression node, VariableStorageKind storage)
	at System.Linq.Expressions.Compiler.VariableBinder.VisitParameter(ParameterExpression node)
	at System.Linq.Expressions.ParameterExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
	at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
	at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
	at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
	at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
	at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
	at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
	at System.Linq.Expressions.ExpressionVisitor.VisitNewArray(NewArrayExpression node)
	at System.Linq.Expressions.NewArrayExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
	at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
	at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
	at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
	at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
	at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
	at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
	at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
	at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
	at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
	at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
	at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
	at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda, DebugInfoGenerator debugInfoGenerator)
	at System.Linq.Expressions.Expression`1.Compile()
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateExecutorLambda[TResults]()
	at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.CreateExecutorLambda[TResults]()
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateAsyncQueryExecutor[TResult](QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Storage.Database.CompileAsyncQuery[TResult](QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileAsyncQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass22_0`1.<CompileAsyncQuery>b__0()
	at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
	at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddAsyncQuery[TResult](Object cacheKey, Func`1 compiler)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileAsyncQuery[TResult](Expression query)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query)
	at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression)
	at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.System.Collections.Generic.IAsyncEnumerable<TResult>.GetEnumerator()
```

query plan:

```
(QueryContext queryContext) => IAsyncEnumerable<<>f__AnonymousType234<Level1, ICollection<Level2>>> _InterceptExceptions(
|__ source: IAsyncEnumerable<<>f__AnonymousType234<Level1, ICollection<Level2>>> _SelectAsync(
|   |__ source: IAsyncEnumerable<Level1> _ShapedQuery(
|   |   |__ queryContext: queryContext, 
|   |   |__ shaperCommandContext: SelectExpression: 
|   |   |       SELECT [l].[Id], [l].[Date], [l].[Name], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_SelfId]
|   |   |       FROM [LevelOne] AS [l]
|   |   |       ORDER BY [l].[Id], 
|   |   |__ shaper: BufferedEntityShaper<Level1>), 
|   |__ selector: (Level1 l | CancellationToken ct) => Task<<>f__AnonymousType234<Level1, ICollection<Level2>>> _ExecuteAsync(
|       |__ taskFactories: new Func<Task<object>>[]
|       |   { 
|       |       () => Task<object> _ToObjectTask(Task<Level1> _IncludeAsync(
|       |       |__ queryContext: queryContext, 
|       |       |__ entity: l, 
|       |       |__ included: new object[]{ }, 
|       |       |__ fixup: (QueryContext queryContext | Level1 entity | object[] included | CancellationToken ct) => 
|       |       |   {
|       |       |       return Task queryContext.QueryBuffer.IncludeCollectionAsync(
|       |       |       |__ includeId: 0, 
|       |       |       |__ navigation: Level1.OneToMany_Optional, 
|       |       |       |__ inverseNavigation: Level2.OneToMany_Optional_Inverse, 
|       |       |       |__ targetEntityType: EntityType: Level2, 
|       |       |       |__ clrCollectionAccessor: ClrICollectionAccessor<Level1, ICollection<Level2>, Level2>, 
|       |       |       |__ inverseClrPropertySetter: ClrPropertySetter<Level2, Level1>, 
|       |       |       |__ tracking: False, 
|       |       |       |__ instance: entity, 
|       |       |       |__ valuesFactory: (Func<IAsyncEnumerable<Level2>>)() => IAsyncEnumerable<Level2> _ShapedQuery(
|       |       |       |   |__ queryContext: queryContext, 
|       |       |       |   |__ shaperCommandContext: SelectExpression: 
|       |       |       |   |       SELECT [l.OneToMany_Optional].[Id], [l.OneToMany_Optional].[Date], [l.OneToMany_Optional].[Level1_Optional_Id], [l.OneToMany_Optional].[Level1_Required_Id], [l.OneToMany_Optional].[Name], [l.OneToMany_Optional].[OneToMany_Optional_InverseId], [l.OneToMany_Optional].[OneToMany_Optional_Self_InverseId], [l.OneToMany_Optional].[OneToMany_Required_InverseId], [l.OneToMany_Optional].[OneToMany_Required_Self_InverseId], [l.OneToMany_Optional].[OneToOne_Optional_PK_InverseId], [l.OneToMany_Optional].[OneToOne_Optional_SelfId]
|       |       |       |   |       FROM [LevelTwo] AS [l.OneToMany_Optional]
|       |       |       |   |       INNER JOIN (
|       |       |       |   |           SELECT [l0].[Id]
|       |       |       |   |           FROM [LevelOne] AS [l0]
|       |       |       |   |       ) AS [t] ON [l.OneToMany_Optional].[OneToMany_Optional_InverseId] = [t].[Id]
|       |       |       |   |       ORDER BY [t].[Id], 
|       |       |       |   |__ shaper: BufferedEntityShaper<Level2>), 
|       |       |       |__ joinPredicate: default(Func<Level1, Level2, bool>), 
|       |       |       |__ cancellationToken: ct)
|       |       |   }, 
|       |       |__ cancellationToken: Unhandled parameter: ct)), 
|       |       () => Task<object> _ToObjectTask(Task<ICollection<Level2>> Unhandled parameter: queryContext.QueryBuffer.CorrelateSubqueryAsync(
|       |       |__ correlatedCollectionId: 0, 
|       |       |__ navigation: Level1.OneToMany_Optional, 
|       |       |__ resultCollectionFactory: (INavigation n) => (ICollection<Level2>)object IClrCollectionAccessor GetCollectionAccessor(n).Create(), 
|       |       |__ outerKey: new MaterializedAnonymousObject(new object[]{ (object)
|       |       |   {
|       |       |       __caller = l
|       |       |       __result = default(Nullable<int>)
|       |       |       __caller != default(Level1) ? __result = Nullable<int> GetValueFromEntity(
|       |       |           clrPropertyGetter: ClrPropertyGetter<Level1, int>, 
|       |       |           entity: __caller) : default(void)
|       |       |       return __result
|       |       |   } }), 
|       |       |__ tracking: False, 
|       |       |__ correlatedCollectionFactory: (Func<IAsyncEnumerable<Tuple<Level2, MaterializedAnonymousObject, MaterializedAnonymousObject>>>)() => IAsyncEnumerable<Tuple<Level2, MaterializedAnonymousObject, MaterializedAnonymousObject>> _ShapedQuery(
|       |       |   |__ queryContext: Unhandled parameter: queryContext, 
|       |       |   |__ shaperCommandContext: SelectExpression: 
|       |       |   |       SELECT [l.OneToMany_Optional0].[Id], [l.OneToMany_Optional0].[Date], [l.OneToMany_Optional0].[Level1_Optional_Id], [l.OneToMany_Optional0].[Level1_Required_Id], [l.OneToMany_Optional0].[Name], [l.OneToMany_Optional0].[OneToMany_Optional_InverseId], [l.OneToMany_Optional0].[OneToMany_Optional_Self_InverseId], [l.OneToMany_Optional0].[OneToMany_Required_InverseId], [l.OneToMany_Optional0].[OneToMany_Required_Self_InverseId], [l.OneToMany_Optional0].[OneToOne_Optional_PK_InverseId], [l.OneToMany_Optional0].[OneToOne_Optional_SelfId], [t0].[Id]
|       |       |   |       FROM [LevelTwo] AS [l.OneToMany_Optional0]
|       |       |   |       INNER JOIN (
|       |       |   |           SELECT [l1].[Id]
|       |       |   |           FROM [LevelOne] AS [l1]
|       |       |   |       ) AS [t0] ON [l.OneToMany_Optional0].[OneToMany_Optional_InverseId] = [t0].[Id]
|       |       |   |       ORDER BY [t0].[Id], 
|       |       |   |__ shaper: (QueryContext queryContext | TransparentIdentifier<Level2, MaterializedAnonymousObject> t0) => new Tuple<Level2, MaterializedAnonymousObject, MaterializedAnonymousObject>(
|       |       |           t0.Outer, 
|       |       |           new MaterializedAnonymousObject(new object[]{ (object)
|       |       |           {
|       |       |               __caller = t0.Outer
|       |       |               __result = default(Nullable<int>)
|       |       |               __caller != default(Level2) ? __result = Nullable<int> GetValue(
|       |       |               |__ queryContext: queryContext, 
|       |       |               |__ entity: __caller, 
|       |       |               |__ property: Level2.OneToMany_Optional_InverseId) : default(void)
|       |       |               return __result
|       |       |           } }), 
|       |       |           new MaterializedAnonymousObject(new object[]{ (object)object t0.Inner.GetValue(0) })
|       |       |       )), 
|       |       |__ correlationPredicate: (MaterializedAnonymousObject o | MaterializedAnonymousObject i) => object i.GetValue(0) == default(object) || object o.GetValue(0) == default(object) ? False : (Nullable<int>)(int)object o.GetValue(0) == (Nullable<int>)object i.GetValue(0), 
|       |       |__ cancellationToken: Unhandled parameter: ct)) 
|       |   }, 
|       |__ selector: (object[] results) => new { 
|               (Level1)results[0], 
|               (ICollection<Level2>)results[1]
|            })), 
|__ contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
|__ logger: DiagnosticsLogger<Query>, 
|__ queryContext: Unhandled parameter: queryContext)
```

**Comment by maumar** [comment]
possibly related to #12358

**Comment by smitpatel** [comment]
Perhaps this `ct` is not replaced with `queryContext.CancellationTokenParameter`

**Comment by maumar** [comment]
fixed in 3d2ba391cc1d8d6cc7b0a67ced4539067ad07a64
